tests-ng/DEVELOPER.md

```diff
@@ -0,0 +1,490 @@
+- **Examples**: `equals_ignore_case`, `parse_etc_file`
+
+```python
+# Example utility usage in plugin
+def get_config_section(self, section: str):
+    # Uses utils internally
+    return parse_etc_file(self.config_path, [section])
+```
+
+## Markers and Test Configuration
+
+### Core Markers
+
+#### `@pytest.mark.booted(reason="...")`
+
+Indicates the test requires a booted system:
+
+```python
+@pytest.mark.booted(reason="Calling sshd -T requires a booted system")
+def test_sshd_config(sshd: Sshd):
+    # Test implementation
+```
+
+#### `@pytest.mark.modify(reason="...")`
+
+Indicates the test modifies system state:
+
+```python
+@pytest.mark.modify(reason="Starting the unit modifies the system state")
+def test_service_start(systemd: Systemd):
+    # Test implementation
+```
+
+#### `@pytest.mark.root(reason="...")`
+
+Indicates the test requires root privileges:
+
+```python
+@pytest.mark.root(reason="Starting the unit requires root")
+def test_service_start(systemd: Systemd):
+    # Test implementation
+```
+
+#### `@pytest.mark.feature("condition", reason="...")`
+
+Limits test execution based on feature conditions:
+
+```python
+@pytest.mark.feature("ssh and not (ali or aws or azure or openstack)",
+                     reason="We want no authorized_keys for unmanaged users")
+def test_users_have_no_authorized_keys(expected_users):
+    # Test implementation
+```
+
+#### `@pytest.mark.performance_metric`
+
+Marks performance tests that can be skipped under emulation:
+
+```python
+@pytest.mark.performance_metric
+@pytest.mark.booted(reason="We can only measure startup time if we actually boot the system")
+def test_startup_time(systemd: Systemd):
+    # Test implementation
+```
+
+### Parametrization
+
+Use `@pytest.mark.parametrize` for testing multiple scenarios:
+
+```python
+@pytest.mark.parametrize("sshd_config_item", required_sshd_config)
+def test_sshd_has_required_config(sshd_config_item: str, sshd: Sshd):
+    actual_value = sshd.get_config_section(sshd_config_item)
+    expected_value = required_sshd_config[sshd_config_item]
+    # Test implementation
+```
+
+### Missing Markers (TODO)
+
+#### `@pytest.mark.security_id`
+
+TODO: Explain what this marker is good for.
+
+## Python Best Practices
+
+### Code Style and CI Enforcement
+
+The project enforces code quality through CI linting (see `.github/workflows/lint_tests-ng.yml`):
+
+- **[Black](https://black.readthedocs.io/en/stable/)**: Automatic code formatting
+- **[isort](https://pycqa.github.io/isort/)**: Import statement sorting
+- **[Pyright](https://github.com/microsoft/pyright/blob/main/docs/getting-started.md)**: Static type checking
+
+**Guidelines:**
+
+- Follow [PEP 8 – Style Guide for Python Code](https://peps.python.org/pep-0008/)
+- Use type hints for function parameters and return values
+- Prefer descriptive variable names
+- Use docstrings for complex functions
+
+> [!TIP]
+> Run `make -f tests-ng/dev.makefile format` locally before committing
```

Comment:
I'd be pretty much in favor to introduce `.pre-commit` to help with that. Open for discussions.